File: src/main/java/com/github/bzumhagen/sct/ChangelogChange.java
package com.github.bzumhagen.sct;

import com.github.zafarkhaja.semver.Version;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

final class ChangelogChange {
    private String description;
    private Version version;
    private String type;
    private String reference;

    ChangelogChange(String description, Version version, String type, String reference) {
        this.description = description;
        this.version = version;
        this.type = type;
        this.reference = reference;
    }

    public String getDescription() {
        return description;
    }

    public Version getVersion() {
        return version;
    }

    public String getType() {
        return type;
    }

    public String getReference() { return reference; }

    @Override
    public int hashCode() {
        return new HashCodeBuilder(17, 31). // two randomly chosen prime numbers
                // if deriving: appendSuper(super.hashCode()).
                        append(description).
                        append(version).
                        append(type).
                        append(reference).
                        toHashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof ChangelogChange))
            return false;
        if (obj == this)
            return true;

        ChangelogChange rhs = (ChangelogChange) obj;
        return new EqualsBuilder().
                // if deriving: appendSuper(super.equals(obj)).
                        append(description, rhs.description).
                        append(version, rhs.version).
                        append(type, rhs.type).
                        append(reference, rhs.reference).
                        isEquals();
    }
}

File: src/main/scala/com/github/bzumhagen/sct/ChangelogChange.scala
package com.github.bzumhagen.sct

import java.time.LocalDate

import com.github.zafarkhaja.semver.Version

/** A changelog change
  *
  * @param description change description
  * @param version change version
  * @param changeType change type (i.e. Added)
  * @param reference change reference (i.e. XYZ-123)
  * @param date change date
  */
object ChangelogChange {
  def apply(description: String, version: Version, changeType: String, date: LocalDate, reference: Option[String] = None) =
    new ChangelogChange(description, version, changeType, reference.map(ChangelogReference), date)
}
case class ChangelogChange(description: String, version: Version, changeType: String, reference: Option[ChangelogReference], date: LocalDate)

case class ChangelogReference(value: String)